### PR TITLE
Effect plugin window beautifications

### PR DESCRIPTION
**Pull request by ksoze95**
Resolves: #3090 

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/EffectUI.cpp [hunk 1]
```diff
@@ -309,15 +309,24 @@ int EffectUIHost::ShowModal()
    mApplyBtn->Destroy();
    mApplyBtn = apply;
    mApplyBtn->SetDefault();
-   mApplyBtn->SetLabel(wxGetStockLabel(wxID_OK, 0));
+
+   if (mIsBatch)
+      mApplyBtn->SetLabel(wxGetStockLabel(wxID_OK, 0));
+   else if (mEffectUIHost.GetDefinition().GetType() == EffectTypeGenerate)
+      mApplyBtn->SetLabel(XO("Generate").Translation());
```
Comment (crsib) [inline src/effects/EffectUI.cpp]: I would like to see `XXO("&Generate")`. This way this string will be reused.

File: src/effects/EffectUI.cpp [hunk 2]
```diff
@@ -309,15 +309,24 @@ int EffectUIHost::ShowModal()
    mApplyBtn->Destroy();
    mApplyBtn = apply;
    mApplyBtn->SetDefault();
-   mApplyBtn->SetLabel(wxGetStockLabel(wxID_OK, 0));
+
+   if (mIsBatch)
```
Comment (crsib) [inline src/effects/EffectUI.cpp:302]: Probably a wrong condition?

File: src/effects/EffectUI.cpp [hunk 3]
```diff
@@ -309,15 +309,24 @@ int EffectUIHost::ShowModal()
...
+   else if (mEffectUIHost.GetDefinition().GetType() == EffectTypeGenerate)
+      mApplyBtn->SetLabel(XO("Generate").Translation());
+
    mCloseBtn->SetLabel(wxGetStockLabel(wxID_CANCEL, 0));
 #else
-   mApplyBtn->SetLabel(wxGetStockLabel(wxID_OK));
+   if (mIsBatch)
+      mApplyBtn->SetLabel(wxGetStockLabel(wxID_OK));
+   else if (mEffectUIHost.GetDefinition().GetType() == EffectTypeGenerate)
+      mApplyBtn->SetLabel(XO("Generate").Translation());
```
Comment (crsib) [inline src/effects/EffectUI.cpp:312]: Ditto

File: src/effects/EffectUI.cpp [hunk 4]
```diff
@@ -357,73 +344,46 @@ wxPanel *EffectUIHost::BuildButtonBar(wxWindow *parent, bool graphicalUI)
...
-            {
-               mPlayToggleBtn = S.Id( kPlayID )
-                  .ToolTip(XO("Start and stop playback"))
-                  .AddButton( XXO("Start &Playback"),
-                              wxALIGN_CENTER | wxTOP | wxBOTTOM );
-            }
-            else if (
-               (mEffectUIHost.GetDefinition().GetType() != EffectTypeAnalyze) &&
-               (mEffectUIHost.GetDefinition().GetType() != EffectTypeTool) )
+            if (mwState.expired())
```
Comment (crsib) [inline src/effects/EffectUI.cpp:367]: Wrong condition? After the #3391 mwState should not be used.

File: src/effects/EffectUI.cpp [hunk 5]
```diff
@@ -357,73 +344,46 @@ wxPanel *EffectUIHost::BuildButtonBar(wxWindow *parent, bool graphicalUI)
...
             {
-               mPlayToggleBtn = S.Id( kPlayID )
-                  .ToolTip(XO("Preview effect"))
-                  .AddButton( XXO("&Preview"),
+               S.AddSpace( 5, 5 );
+
+               mPlayToggleBtn = S.Id(kPlayID)
+                  .ToolTip(XO("Start and stop preview"))
+                  .AddButton( XXO("Start &Preview"),
```
Comment (crsib) [inline src/effects/EffectUI.cpp:373]: I have checked all the design sources - button is always called "Preview". Also, it should be "grouped" with the "Apply", right now it leans to the left (i. e. it is near the "Presets" button)

File: src/effects/RealtimeEffectState.h [hunk 6]
```diff
@@ -87,6 +91,9 @@ class RealtimeEffectState
    //! Main thread cleans up playback
    bool Finalize() noexcept;
 
+   //! Notify all subscribers of a state change
+   void PublishStateUpdate(RealtimeEffectStateChange state);
```
Comment (crsib) [inline src/effects/RealtimeEffectState.h:95]: You can do `using Observer::Publisher<RealtimeEffectStateChange>::Publish`, but I don't yet understand why this function is public

File: src/RealtimeEffectPanel.cpp [hunk 7]
```diff
@@ -458,6 +469,25 @@ namespace
          effectStateUI.Toggle( *mProject );
       }
 
+      void OnEnableButtonClicked(wxCommandEvent& event)
+      {
+         auto pButton = static_cast<ThemedButtonWrapper<wxBitmapButton>*>(mEnableButton);
+         auto index = pButton->GetBitmapIndex();
+         bool wasEnabled = (index == bmpEffectOn);
+
+         if (mSettingsAccess) {
```
Comment (crsib) [inline src/RealtimeEffectPanel.cpp:478]: I think this should become a `RealtimeEffectState::SetActive`. That specific method should also Publish the message

**Review by crsib — CHANGES_REQUESTED** [review]
* Opening fancy UI from the RT panel crashes Audacity
* Using Beat Finder crashes
* Preview button is in a wrong place
* Some other feedback in comments

Also I want a design review to be completed first. "Non fancy" UI looks quite odd, especially with built-in generators. Probably we need a horizontal line to visually separate the buttons from effect settings.

File: src/effects/EffectUI.cpp (hunk 5 above)
Comment (LWinterberg) [reply src/effects/EffectUI.cpp:373]: Agree on calling it "preview" (changing to "stop preview" after it's been pressed). Regarding the grouping, it apparently was difficult and martin didn't mind it ungrouped iirc.

Comment (crsib) [reply src/effects/EffectUI.cpp:373]: Well, I know how to do such grouping :-)

File: src/effects/RealtimeEffectState.h (hunk 6 above)
Comment (ksoze95) [reply src/effects/RealtimeEffectState.h:95]: Its also called inside EffectUIHost::OnEnable to signal back to the realtime side panel that there has been a state change.

File: src/effects/EffectUI.cpp (hunk 1 above)
Comment (ksoze95) [reply src/effects/EffectUI.cpp]: Done.

File: src/effects/EffectUI.cpp (hunk 3 above)
Comment (ksoze95) [reply src/effects/EffectUI.cpp]: Changed. to XXO("&Generate").

File: src/effects/RealtimeEffectState.h (hunk 6 above)
Comment (ksoze95) [reply src/effects/RealtimeEffectState.h]: Done.

File: src/RealtimeEffectPanel.cpp (hunk 7 above)
Comment (ksoze95) [reply src/RealtimeEffectPanel.cpp]: Done.

File: src/effects/EffectUI.cpp (hunk 2 above)
Comment (ksoze95) [reply src/effects/EffectUI.cpp]: Changed to use mpTempProjectState instead, plus tacked on a few other condition checks as well.

File: src/effects/RealtimeEffectState.cpp [hunk 8]
```diff
@@ -545,6 +545,20 @@ bool RealtimeEffectState::IsActive() const noexcept
    return mWorkerSettings.extra.GetActive();
 }
 
+void RealtimeEffectState::SetActive(bool active)
+{
+   if (auto access = GetAccess()) {
```
Comment (vsverchinsky) [inline src/effects/RealtimeEffectState.cpp:550]: `GetAccess` does not return empty pointer

File: src/effects/RealtimeEffectState.h [hunk 9]
```diff
@@ -20,17 +20,21 @@
...
 #include "GlobalVariable.h"
 #include "MemoryX.h"
+#include "Observer.h"
 #include "PluginProvider.h" // for PluginID
 #include "XMLTagHandler.h"
 
 class EffectSettingsAccess;
 class Track;
 
+enum class RealtimeEffectStateChange : char { EffectOff, EffectOn };
```
Comment (crsib) [inline src/effects/RealtimeEffectState.h:30]: Do we really need `char` here?

File: src/effects/EffectUI.cpp [hunk 10]
```diff
@@ -1278,17 +1113,18 @@ std::shared_ptr<EffectInstance> EffectUIHost::InitializeInstance()
...
    return result;
 }
 
 void EffectUIHost::CleanupRealtime()
 {
-   mSubscription.Reset();
+   mAudioIOSubscription.Reset();
 
    if (mSupportsRealtime && mInitialized) {
       if (mpTempProjectState) {
```
Comment (crsib) [inline src/effects/EffectUI.cpp:1125]: That not yet created method here please as well

**Review by crsib — CHANGES_REQUESTED** [review]
There is still some feedback to addressed left. We can waive that enable button bug for now, I don't want it to block this PR.

(But we need to log that bug)

File: src/effects/EffectUI.cpp (hunk 4 above)
Comment (ksoze95) [reply src/effects/EffectUI.cpp]: Done.

File: src/effects/EffectUI.cpp [hunk 11]
```diff
@@ -355,111 +339,44 @@ wxPanel *EffectUIHost::BuildButtonBar(wxWindow *parent, bool graphicalUI)
...
       S.SetBorder( margin );
 
-      if (!mIsGUI)
-      {
-         mMenuBtn = S.Id( kMenuID )
-            .ToolTip(XO("Manage presets and options"))
-            .AddButton( XXO("&Manage"), wxALIGN_CENTER | wxTOP | wxBOTTOM );
-      }
-      else
+      if (!mpTempProjectState && mSupportsRealtime)
```
Comment (crsib) [inline src/effects/EffectUI.cpp:342]: `IsOpenedFromEffectPanel`? Minor though

File: src/effects/EffectUI.cpp [hunk 12]
```diff
@@ -355,111 +339,44 @@ wxPanel *EffectUIHost::BuildButtonBar(wxWindow *parent, bool graphicalUI)
...
-                  .ToolTip(XO("Skip forward"))
-                  .Name(XO("Skip &Forward"))
-                  .AddBitmapButton( CreateBitmap(
-                     effect_ffwd_xpm, true, true) );
-               mFFwdBtn->SetBitmapDisabled(
-                  CreateBitmap(effect_ffwd_disabled_xpm, true, false));
-               mFFwdBtn->SetBitmapPressed(CreateBitmap(effect_ffwd_xpm, false, true));
-            }
-
             S.AddSpace( 5, 5 );
```
Comment (crsib) [inline src/effects/EffectUI.cpp:374]: It appears that both branches are identical now

**Review by crsib — APPROVED** [review]
I think I can approve it now, with a couple minorish problems. Please rebase the branch on master first

**Pull request merged**